parts/linux/cloud-init/artifacts/ubuntu/ubuntu-snapshot-update.sh

```diff
@@ -79,11 +79,41 @@ if [ -n "${current_timestamp}" ]; then
 fi
 
 old_source_list=$(cat ${source_list_path})
-# upgrade from base image to a timestamp
-# e.g. replace https://snapshot.ubuntu.com/ubuntu/ with https://snapshot.ubuntu.com/ubuntu/20230727T000000Z
-sed -i 's/http:\/\/azure.archive.ubuntu.com\/ubuntu\//https:\/\/snapshot.ubuntu.com\/ubuntu\/'"${golden_timestamp}"'/g' ${source_list_path}
-# upgrade from one timestamp to another timestamp
-sed -i 's/https:\/\/snapshot.ubuntu.com\/ubuntu\/\([0-9]\{8\}T[0-9]\{6\}Z\)/https:\/\/snapshot.ubuntu.com\/ubuntu\/'"${golden_timestamp}"'/g' ${source_list_path}
+
+# Network isolated cluster can't access the internet, so we deploy a live patching repo service in the cluster
+# The node will use the live patching repo service to download the repo metadata and packages
+# If the annotation is not set, we will use the ubuntu snapshot repo
+live_patching_repo_service=$($KUBECTL get node ${node_name} -o jsonpath="{.metadata.annotations['kubernetes\.azure\.com/live-patching-repo-service']}")
+# Limit the live patching repo service to private IPs in the range of 10.x.x.x, 172.16.x.x - 172.31.x.x, and 192.168.x.x
+private_ip_regex="^((10\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})|(172\.(1[6-9]|2[0-9]|3[01])\.[0-9]{1,3}\.[0-9]{1,3})|(192\.168\.[0-9]{1,3}\.[0-9]{1,3}))$"
+if [[ -n "${live_patching_repo_service}" && ! "${live_patching_repo_service}" =~ $private_ip_regex ]]; then
+    echo "Ignore invalid live patching repo service: ${live_patching_repo_service}"
+    live_patching_repo_service=""
+fi
+if [ -z "${live_patching_repo_service}" ]; then
+    echo "live patching repo service is not set, use ubuntu snapshot repo"
+    # upgrade from base image to a timestamp
+    # e.g. replace https://azure.archive.ubuntu.com/ubuntu/ with https://snapshot.ubuntu.com/ubuntu/20230727T000000Z
+    sed -i 's/http:\/\/azure.archive.ubuntu.com\/ubuntu\//https:\/\/snapshot.ubuntu.com\/ubuntu\/'"${golden_timestamp}"'/g' ${source_list_path}
+    # upgrade from one timestamp to another timestamp
+    # e.g. replace https://snapshot.ubuntu.com/ubuntu/20250310T000000Z with https://snapshot.ubuntu.com/ubuntu/20250318T000000Z
+    sed -i 's/https:\/\/snapshot.ubuntu.com\/ubuntu\/\([0-9]\{8\}T[0-9]\{6\}Z\)/https:\/\/snapshot.ubuntu.com\/ubuntu\/'"${golden_timestamp}"'/g' ${source_list_path}
+    # No live patching repo service annotation, so we need to change to use the ubuntu snapshot repo
+    # e.g. replace http://10.224.0.5/ubuntu/ with https://snapshot.ubuntu.com/ubuntu/20250318T000000Z
+    sed -i 's/http:\/\/[0-9]\+.[0-9]\+.[0-9]\+.[0-9]\+\/ubuntu\//https:\/\/snapshot.ubuntu.com\/ubuntu\/'"${golden_timestamp}"'/g' ${source_list_path}
```

Comment:
not related to this PR, but we need to make the same changes to the cshelper in RP

Reply:
Yes, that's what we need to update in captureVHD API if it's NI:

1. Get service IP
2. Update cse for copied VMSS